spec.bs

```diff
@@ -718,50 +779,37 @@ To <dfn>generate and score bids</dfn> given an [=auction config=] |auctionConfig
-          [=interest group/bidding url=].
-        1. If |biddingScript| is failure, [=iteration/continue=].
-        1. If |ig|'s [=interest group/bidding wasm helper url=] is not null:
-          1. Let |wasmModuleObject| be the result of [=fetching WebAssembly=] with |ig|'s
-            [=interest group/bidding wasm helper url=].
-          1. If |wasmModuleObject| is not failure:
-            1. [=map/Set=] |browserSignals|["`wasmHelper`"] to |wasmModuleObject|.
-        1. Let |trustedBiddingSignals| be an [=ordered map=] whose [=map/keys=] are [=strings=] and
-          whose [=map/values=] are {{any}}.
-        1. [=list/For each=] |key| of |ig|'s [=interest group/trusted bidding signals keys=]:
-          1. If |allTrustedBiddingSignals| is an [=ordered map=] and |allTrustedBiddingSignals|[|key|]
-            [=map/exists=], then [=map/set=] |trustedBiddingSignals|[|key|] to
-            |allTrustedBiddingSignals|[|key|].
-        1. Let |startTime| be the current time.
-        1. Let |generatedBidResult| be the result of [=evaluating a bidding script=] with
-           |biddingScript|, |ig|, « |interestGroup|, |auctionSignals|, |perBuyerSignals|,
-           |trustedBiddingSignals|, |browserSignals| », and |perBuyerTimeout|.
-        1. Let |duration| be the current time minus |startTime| in milliseconds.
-        1. If |generatedBidResult| is an [=abrupt completion=], [=iteration/continue=].
-        1. Let |generatedBidIDL| be the result of [=converted to an IDL value|converting=]
-          |generatedBidResult| to a {{GenerateBidOutput}}.
-        1. If an exception was caught in the previous step, [=iteration/continue=].
-        1. Let |groupHasAdComponents| be true.
-        1. If |interestGroup|'s [=interest group/ad components=] is null:
-          1. Set |groupHasAdComponents| be false.
-        1. Let |generatedBid| be the result of [=converting GenerateBidOutput to generated bid=]
-          with |generatedBidIDL|, |ig|, |isComponentAuction|, and |groupHasAdComponents|.
-        1. Set |generatedBid|'s [=generated bid/bid duration=] to |duration|.
+        1. Let |generatedBid| be the result of [=generate a bid=] given |isComponentAuction|,
+          |allTrustedBiddingSignals|, |auctionSignals|, a [=map/clone=] of |browserSignals|,
+          |perBuyerSignals|, |perBuyerTimeout|, and |ig|.
         1. If |generatedBid| is failure, [=iteration/continue=].
-        1. Set |generatedBid|'s [=generated bid/interest group=] to |ig|.
+        1. If [=query generated bid k-anonymity count=] given |generatedBid| returns false:
+          Note: [=Generate a bid=] is now rerun with only k-anonymous [=interest group/ads=] to give
+          the buyer a chance to [=generate a bid=] for k-anonymous [=interest group/ads=]. Allowing
+          the buyer to first [=generate a bid=] for non-k-anonymous [=interest group/ads=] provides a
+          mechanism to bootstrap the k-anonymity count, otherwise no [=interest group/ads=] would

```

Comment:
I don't see any calls to "increment ad k-anonymity count" that could see the first ad. Incrementing the k-anonymity counts happens in https://pr-preview.s3.amazonaws.com/WICG/turtledove/pull/609.html#dom-navigator-runadauction, which calls this whole algorithm as a unit and so can't see the internal value. Should you move that call into here?

Reply:
Ya, I realized last week when I was half way through this PR that it was missing a good sized chunk to address this, namely:
1. score non-k-anon ads
2. keep track of highest scoring non-k-anon ad
3. after auction, if highest scoring non-k-anon ad has a higher score than the k-anon one, then increment it's k-anon count.

This PR was already pretty big so I was going to do this in a follow-up PR.  I'll add a TODO to this one so it doesn't get forgotten.

Reply:
Sounds good.